src/heretic/sb_bar.c

```diff
@@ -1300,33 +1356,304 @@ static void CheatShowFpsFunc(player_t* player, Cheat_t* cheat)
+        if (player->weaponowned[i])
+        {
+            return i;
+        }
+    }
+
+    if (player->weaponowned[wp_gauntlets])
+    {
+        return wp_gauntlets;
+    }
+
+    return wp_staff;
+}
+
+static void CheatAddRemoveWpnFunc(player_t *player, Cheat_t *cheat)
+{
+    int i, w, setmsg;
+    char arg;
+
     NIGHTMARE_NETGAME_CHECK;
+
+    // don't do anything if currently chicken
     if (player->chickenTics)
     {
         return;
     }
-    for (i = 1; i < 8; i++)
+
+    cht_GetParam(cheat->seq, &arg);
+    w = arg-'1';
+
+    // WEAP0 takes away all weapons and ammo except for the elvenwand
+    // and 50 crystal ammo
+    if (w == -1)
     {
-        player->weaponowned[i] = false;
+        // remove backpack if the player has it
+        if (player->backpack)
+        {
+            player->backpack = false;
+            for (i = 0; i < NUMAMMO; i++)
+            {
+                player->maxammo[i] /= 2;
+            }
+        }
+
+        // remove Tome of Power if active
+        player->powers[pw_weaponlevel2] = 0;
+
+        for (i = 0; i < NUMWEAPONS; i++)
+        {
+            player->weaponowned[i] = false;
+        }
+        player->weaponowned[wp_staff] = true;
+        player->weaponowned[wp_goldwand] = true;
+
+        for (i = 0; i < NUMAMMO; i++)
+        {
+            player->ammo[i] = 0;
+        }
+        player->ammo[am_goldwand] = 50; // not HHE modifiable?
+
+        if (player->readyweapon > wp_goldwand)
+        {
+            player->pendingweapon = wp_goldwand;
+        }
+
+        P_SetMessage(player, DEH_String(TXT_CHEATWEAPREMOVEALL), false);
+
+        return;
+    }
+
+    setmsg = 0;
+
+    // do not give registered only weapons if in shareware mode
+    // or touch the chicken beak, never remove staff
+    if (!WeaponAvailable(w) || w == wp_staff)
+    {
+        return;
+    }
+
+    if (!player->weaponowned[w])
+    {
+        P_GiveWeapon(player, w);
+        S_StartSound(NULL, sfx_wpnup);
+
+        // no pickup message for staff or wand
+        if (w > 1)
+        {
+            P_SetMessage(player, DEH_String(WeaponPickupMessages[w]), false);
+            setmsg = 1;
+        }
+
+        // trigger palette flash
+        player->bonuscount += 6; // same as BONUSADD
+    }
+    else
+    {
+        player->weaponowned[w] = false;
+
+        // if current weapon was removed, select another one
+        if (w == player->readyweapon)
+        {
+            player->pendingweapon = PickBestWeapon(player);
+        }
+    }
+
+    if (!setmsg)
+    {
+        DEH_snprintf(msgbuf, sizeof(msgbuf),
+         player->weaponowned[w] ? TXT_CHEATWEAPADD : TXT_CHEATWEAPREMOVE,
+         w+1);
+        P_SetMessage(player, msgbuf, false);
     }
-    player->pendingweapon = wp_staff;
-    P_SetMessage(player, DEH_String(TXT_CHEATIDKFA), true);
 }
 
-static void CheatIDDQDFunc(player_t * player, Cheat_t * cheat)
+// [crispy] trigger all special lines available on the map
+#define NO_INDEX (-1)
```

Comment:
Let's please move this to `r_def.h` where it belongs. Once we introduce "extended nodes" support into Heretic, the `#define' will be already there and needs to be adjusted in only one place.